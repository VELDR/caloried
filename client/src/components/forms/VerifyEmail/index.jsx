import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';
import { connect, useDispatch } from 'react-redux';
import OTPInput from 'react-otp-input';
import Countdown from 'react-countdown';
import Swal from 'sweetalert2';
import { createStructuredSelector } from 'reselect';
import mail from '@static/images/mail-sent.svg';

import { selectAccount } from '@pages/SignUp/selectors';
import { resetForm } from '@pages/SignUp/actions';
import { resendVerificationEmail, verifyOTP } from '@containers/Client/actions';

import classes from './style.module.scss';

const VerifyEmail = ({ account, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [otp, setOtp] = useState(null);
  const [isResendDisabled, setIsResendDisabled] = useState(true);
  const [timerKey, setTimerKey] = useState(0);
  const [endTime, setEndTime] = useState(Date.now() + 30000);

  const handleResend = () => {
    dispatch(
      resendVerificationEmail({ email: account.email }, () => {
        setIsResendDisabled(true);
        setEndTime(Date.now() + 30000);
        setTimerKey((prevKey) => prevKey + 1);
      })
    );
  };

  const handleSubmitOTP = () => {
    dispatch(
      verifyOTP(otp, account.email, () => {
        Swal.fire({
          title: formatMessage({ id: 'app_verification_success_title' }),
          text: formatMessage({ id: 'app_verification_success_description' }),
          icon: 'success',
          confirmButtonText: formatMessage({ id: 'app_sign_in' }),
          confirmButtonColor: 'var(--color-primary)',
          background: 'var(--color-bg-secondary)',
          customClass: {
            title: classes.swalTitle,
            confirmButton: classes.swalButton,
            popup: classes.swalPopup,
          },
        }).then((result) => {
          if (result.isConfirmed) {
            dispatch(resetForm());
            navigate('/sign-in');
          }
        });
      })
    );
  };

  useEffect(() => {
    setIsResendDisabled(true);
    setEndTime(Date.now() + 30000);
  }, []);

  const renderer = ({ minutes, seconds, completed }) => {
    if (completed && isResendDisabled) {
      setIsResendDisabled(false);
      return <span className={classes.countdown}>00:00</span>;
    }

    return (
      <span className={classes.countdown}>
        {minutes}:{seconds < 10 ? `0${seconds}` : seconds}
      </span>
    );
  };

  return (
    <div className={classes.container} data-testid="verify-email">
      <img src={mail} alt="Email sent illustration" />
      <div className={classes.title}>
        <FormattedMessage id="app_verify_email" />
      </div>
      <div>
        <div className={classes.description}>
          <FormattedMessage id="app_we_sent_email" />
        </div>
        <div className={classes.email}>{account.email}</div>
      </div>
      <div className={classes.otp}>
        <OTPInput
          value={otp}
          onChange={setOtp}
          numInputs={4}
          renderSeparator={<span>-</span>}
          renderInput={(props) => <input {...props} />}
        />
        <div className={classes.button} onClick={handleSubmitOTP}>
          <FormattedMessage id="app_submit" />
        </div>
      </div>
      <div className={classes.description}>
        <span
          className={`${classes.link} ${isResendDisabled ? classes.disabledLink : ''}`}
          onClick={!isResendDisabled ? handleResend : null}
        >
          <FormattedMessage id="app_resend" />
        </span>{' '}
        in <Countdown date={endTime} renderer={renderer} key={timerKey} />
      </div>
    </div>
  );
};

VerifyEmail.propTypes = {
  account: PropTypes.object,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  account: selectAccount,
});

export default injectIntl(connect(mapStateToProps)(VerifyEmail));
