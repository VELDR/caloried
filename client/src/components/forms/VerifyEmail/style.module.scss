@import '@styles/base';

.container {
  @include flex-center-column(0.75rem);
  text-align: center;
  max-width: 75%;
  margin: 0 auto;
  position: relative;
  height: 100%;
  img {
    @include box(8rem);
    position: absolute;
    top: 0;
  }
  .title {
    @include flex-center-column;
    font-size: $font-size-large;
    font-weight: 700;
    margin-top: 6rem;
  }
  .description {
    font-size: $font-size-small;
  }
  .otp {
    @include flex-center-column(1rem);
    input {
      @include input;
    }
    .button {
      @include primary-button;
      font-size: $font-size-small;
      width: fit-content;
    }
  }

  .email {
    font-weight: 700;
    font-size: $font-size-regular;
  }
  .link {
    cursor: pointer;
    font-size: $font-size-small;
    color: var(--color-accent);
    &:hover {
      text-decoration: underline;
    }
  }
}
