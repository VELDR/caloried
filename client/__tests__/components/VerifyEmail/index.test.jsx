import VerifyEmail from '@components/forms/VerifyEmail';
import { render } from '@utils/testHelper';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

beforeEach(() => {
  wrapper = render(<VerifyEmail />);
});

describe('VerifyEmail Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('verify-email')).toBeInTheDocument();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
