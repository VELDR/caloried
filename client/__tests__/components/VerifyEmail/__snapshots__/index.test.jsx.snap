// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmail Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="verify-email"
      >
        <img
          alt="Email sent illustration"
        />
        <div
          class="title"
        >
          Verifikasi Email Anda
        </div>
        <div>
          <div
            class="description"
          >
            Anda hampir sampai! Kami telah mengirim email ke
          </div>
          <div
            class="email"
          />
        </div>
        <div
          class="otp"
        >
          <div
            style="display: flex; align-items: center;"
          >
            <input
              aria-label="Please enter OTP character 1"
              autocomplete="off"
              maxlength="1"
              style="width: 1em; text-align: center;"
              type="text"
              value=""
            />
            <span>
              -
            </span>
            <input
              aria-label="Please enter OTP character 2"
              autocomplete="off"
              maxlength="1"
              style="width: 1em; text-align: center;"
              type="text"
              value=""
            />
            <span>
              -
            </span>
            <input
              aria-label="Please enter OTP character 3"
              autocomplete="off"
              maxlength="1"
              style="width: 1em; text-align: center;"
              type="text"
              value=""
            />
            <span>
              -
            </span>
            <input
              aria-label="Please enter OTP character 4"
              autocomplete="off"
              maxlength="1"
              style="width: 1em; text-align: center;"
              type="text"
              value=""
            />
          </div>
          <div
            class="button"
          >
            Kirim
          </div>
        </div>
        <div
          class="description"
        >
          <span
            class="link disabledLink"
          >
            Kirim Ulang OTP
          </span>
           
          in 
          <span
            class="countdown"
          >
            0
            :
            30
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="verify-email"
    >
      <img
        alt="Email sent illustration"
      />
      <div
        class="title"
      >
        Verifikasi Email Anda
      </div>
      <div>
        <div
          class="description"
        >
          Anda hampir sampai! Kami telah mengirim email ke
        </div>
        <div
          class="email"
        />
      </div>
      <div
        class="otp"
      >
        <div
          style="display: flex; align-items: center;"
        >
          <input
            aria-label="Please enter OTP character 1"
            autocomplete="off"
            maxlength="1"
            style="width: 1em; text-align: center;"
            type="text"
            value=""
          />
          <span>
            -
          </span>
          <input
            aria-label="Please enter OTP character 2"
            autocomplete="off"
            maxlength="1"
            style="width: 1em; text-align: center;"
            type="text"
            value=""
          />
          <span>
            -
          </span>
          <input
            aria-label="Please enter OTP character 3"
            autocomplete="off"
            maxlength="1"
            style="width: 1em; text-align: center;"
            type="text"
            value=""
          />
          <span>
            -
          </span>
          <input
            aria-label="Please enter OTP character 4"
            autocomplete="off"
            maxlength="1"
            style="width: 1em; text-align: center;"
            type="text"
            value=""
          />
        </div>
        <div
          class="button"
        >
          Kirim
        </div>
      </div>
      <div
        class="description"
      >
        <span
          class="link disabledLink"
        >
          Kirim Ulang OTP
        </span>
         
        in 
        <span
          class="countdown"
        >
          0
          :
          30
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
